### ISSUE TITLE
Connection pool leak unless running under debugger

### ISSUE THREAD

**Issue by dsvensson** [jOOQ/jOOQ#14769]
### Expected behavior

Running tests under debugger work the same way as running tests without debugger.

### Actual behavior

Bumped from 3.15.5 to 3.18.0, and tests reliably hang due to empty pool. No good idea how to troubleshoot this. Everything works as expected when running under debugger. The same applies to 3.17.x. What might account for the behavioral difference between debugger vs not debugger? Timing, different execution paths?

### Steps to reproduce the problem

Haven't been able to reproduce this outside of corporate environment.

### jOOQ Version

3.18.0

### Database product and version

PostgreSQL 14.5

### Java Version

17.0.6+10-Ubuntu-0ubuntu122.10

### OS Version

Ubuntu 22.10

### JDBC driver name and version (include name if unofficial driver)

42.5.3

**Comment by lukaseder** [comment]
Thanks a lot for your message.

A Java debugger isn't entirely side effect free. It may instrument the debugged logic, modify it (for inspection), delay it (leading to time based behavioural differences), etc. So, the assumption that behaviour is exactly identical is wrong, although one can reasonably hope that it usually doesn't matter.

I'm not aware of any such issues, so without a specific reproducer, involving all the third parties, it will be hard to help here, I'm afraid. Perhaps you can bisect a bit more? I.e. instead of jumping from 3.15.5 to 3.18.0, try smaller upgrade steps? You said 3.17.x also exposes the behaviour, or doesn't? Can you isolate a specific patch release that introduces the problem?

**Comment by dsvensson** [comment]
Yep, will isolate version.

**Comment by dsvensson** [comment]
* 3.15.x works, currently using this in prod.
* 3.16.0..15 works
* 3.17.0 does not work, running under debugger doesn't help, setting `junit.jupiter.execution.parallel.enabled = false` hides/avoids the connection leak.
* 3.18.0 does not work, running under debuger, or setting `junit.jupiter.execution.parallel.enabled = false` hides/avoids the connection leak.

**Comment by dsvensson** [comment]
Further investigation I configured fixed concurrency level, and limited it to 4 on my 32 core workstation, and tests passed with 3.17.0. Given this, I went on to bump the pool size for the tests, and now the tests reliably pass on 3.17.0. I then got 3.18.0 building again, but pool size had no impact on this setup. The symptoms is that all connections are stuck in ClientRead as wait_event in the pg_activity table. So some kind of concurrency issue. I guess this was the case before as well, but it was masked by connection pool limit being hit.

Standing out in the stacktraces is:

```
    at org.jooq.impl.DSL.using(DSL.java:739)
    - waiting on the Class initialization monitor for org.jooq.impl.DefaultDSLContext
```

It sounds a bit like https://github.com/jOOQ/jOOQ/issues/11200

**Comment by lukaseder** [comment]
Thanks a lot for the detailed analysis. I agree, it always seems to be the same issue causing these things: global defaults being declared in multiple classes, which may (or may not) produce a cyclic `<clinit>` dependency.

The problem is that these instances are initialised eagerly with the class initialisation, rather than lazily. A better solution for the JVM is on the roadmap with a JEP draft:
https://openjdk.org/jeps/8209964

In the meantime, perhaps we should roll our own `Lazy` constant wrapper for all of these. In the past,  the constants were moved around into smaller classes (e.g. nested classes) to make deadlocks less likely, but it can still happen.

**Comment by lukaseder** [comment]
I think the main problem here is really that the `org.jooq.impl.Tools` class is so widely used because it contains all utility functions that don't really need any more specific namespace. It also contains global default instances for `Configuration` and `DSLContext`, if they're absent, and those are the ones we don't want to eagerly initialise just because some random code uses something like `Tools::apply` in a `<clinit>`.

I drafted a `Lazy<T>` utility that delays the initialisation of its "memoized" `Supplier<T>` using double checked locking of a `volatile` member, in order to implement something similar as https://openjdk.org/jeps/8209964 intends to do on a language level:

```java
final class Lazy<T> {

    static <T> Lazy<T> of(Supplier<? extends T> supplier) {
        return new Lazy<>(supplier);
    }

    private final Supplier<? extends T> supplier;
    private volatile T                  value;

    private Lazy(Supplier<? extends T> supplier) {
        this.supplier = supplier;
    }

    T get() {
        if (value == null) {
            synchronized (this) {
                if (value == null) {
                    value = supplier.get();
                }
            }
        }

        return value;
    }
}
```

I'll push this draft shortly to 3.19.0-SNAPSHOT, including a backport to 3.18.1, 3.17.10, 3.16.16. Would love to have your further feedback to verify that this addresses the problem you're observing.

Snapshots can be built directly from github, or downloaded from here, if you're licensed: https://www.jooq.org/download/versions

**Comment by lukaseder** [comment]
Changes have been pushed.

**Comment by lukaseder** [comment]
Well, I'm going to release this today or tomorrow, hoping this will fix your issue.

Fixed in jOOQ 3.19.0, 3.18.1 (https://github.com/jOOQ/jOOQ/issues/14825), and 3.17.10 (https://github.com/jOOQ/jOOQ/issues/14826)

**Comment by lukaseder** [comment]
3.18.1 and 3.17.10 have been released with this attempted fix. Let me know if that solves the problem for you.

**Comment by dsvensson** [comment]
Works!

**Comment by lukaseder** [comment]
Glad to hear that, thanks for the feedback!

**Comment by oleg-smith** [comment]
Hello @lukaseder - is it possible to backport it to a Java 11 compatible version of JOOQ, by chance?

**Comment by lukaseder** [comment]
The commercial distributions aren't available from Maven Central. Please use your license key to download the desired distribution either from the link I've provided or from our commercial Maven repository:

- https://www.jooq.org/download/versions
- https://repo.jooq.org

All commercial editions, including the jOOQ Express Edition give access to the Java 11 distribution, see:
https://www.jooq.org/download/#feature-comparison

**Comment by AndreKoepke** [comment]
I have the same problem with JOOQ (`3.19.8` and `3.19.15`).

**Comment by lukaseder** [comment]
> I have the same problem with JOOQ (`3.19.8` and `3.19.15`).

Please open a new issue with a complete reproducer, knowing 1) similar problems are not necessarly "same problems," and 2) race conditions are very hard to reproduce.
